README.rst
Python Social Auth
==================

.. image:: https://travis-ci.org/emilyhorsman/socialauth.svg?branch=master
    :target: https://travis-ci.org/emilyhorsman/socialauth
    :alt: Travis CI

.. image:: https://coveralls.io/repos/github/emilyhorsman/socialauth/badge.svg?branch=master
    :target: https://coveralls.io/github/emilyhorsman/socialauth?branch=master
    :alt: Coveralls Test Coverage

.. image:: https://readthedocs.org/projects/socialauth/badge/?version=latest
    :target: http://socialauth.readthedocs.org/en/latest/?badge=latest
    :alt: Documentation Status

.. image:: https://codeclimate.com/github/emilyhorsman/socialauth/badges/gpa.svg
    :target: https://codeclimate.com/github/emilyhorsman/socialauth
    :alt: Code Climate

.. image:: https://img.shields.io/pypi/v/socialauth.svg
    :target: https://pypi.python.org/pypi/socialauth
    :alt: PyPI

.. image:: https://img.shields.io/badge/python-%E2%89%A5%203.3-blue.svg
    :target: https://docs.python.org/3/
    :alt: Python >= 3.3

.. image:: https://img.shields.io/badge/code%20of%20conduct-v1.4.0-4C1161.svg
    :target: CODE_OF_CONDUCT.md
    :alt: Contributor Covenant Code of Conduct

.. image:: https://img.shields.io/pypi/l/socialauth.svg
    :alt: MIT License


A library for social sign-in capability from providers such as Twitter and
Facebook.

Many social authentication solutions exist. I wanted something that didn’t
have strong ties to an HTTP framework or storage backend. Preferably, I
didn’t want something that dealt with a storage backend at all. This library
uses JSON web tokens instead of sessions to deal with intermediate information
in the OAuth flow (such as a token secret).

.. code-block:: bash

    $ pip install socialauth



Run Tests
=========

.. code-block:: bash

    $ git clone https://github.com/emilyhorsman/socialauth.git
    $ mkvirtualenv --python=python3 socialauth
    $ cd socialauth
    $ workon .
    $ pip install -r requirements.test.txt
    $ python tests.py

Build Docs
==========

.. code-block:: bash

    $ pip install -r requirements.docs.txt
    $ cd docs
    $ make html



CHANGELOG.rst
0.2.0
=====

See cb14add_

* Changed the result of http_get_provider to give the provider user ID and
  user name.

.. _cb14add: https://github.com/emilyhorsman/socialauth/commit/cb14adde36a0b93724101d0740ded56d197c65a9


tests.py
# -*- coding: utf-8 -*-

import os
import unittest
from unittest.mock import patch
from urllib.parse import urlparse, parse_qs
from contextlib import contextmanager

import jwt
import httplib2
import socialauth


twitter_base_url  = 'https://test.socialauth.foo/auth/twitter'
facebook_base_url = 'https://test.socialauth.foo/auth/facebook'


def mock_request_with(path, status, body = b''):
    def func(self, uri, *args, **kwargs):
        url = urlparse(uri)
        if url[2].endswith(path):
            return (httplib2.Response(dict(status = status)), body)

        return mock_valid_requests(self, uri, *args, **kwargs)

    return func


def mock_valid_requests(self, uri, *args, **kwargs):
    url = urlparse(uri)

    # https://dev.twitter.com/oauth/reference/post/oauth/request_token
    if url[1] == 'api.twitter.com' and url[2] == '/oauth/request_token':
        res = httplib2.Response(dict(status = 200))
        content = b'oauth_token=foo&oauth_token_secret=bar&oauth_callback_confirmed=true'

    # https://dev.twitter.com/oauth/reference/post/oauth/access_token
    if url[1] == 'api.twitter.com' and url[2] == '/oauth/access_token':
        res = httplib2.Response(dict(status = 200))
        content = b'oauth_token=foo&oauth_token_secret=bar&user_id=987&screen_name=test'

    # https://developers.facebook.com/docs/facebook-login/manually-build-a-login-flow#confirm
    if url[1] == 'graph.facebook.com' and url[2].endswith('/oauth/access_token'):
        res = httplib2.Response(dict(status = 200))
        content = b'{"access_token":"foobar","token_type":"bearer","expires_in":5117097}'

    # https://developers.facebook.com/docs/graph-api/using-graph-api
    if url[1] == 'graph.facebook.com' and url[2] == '/me':
        res = httplib2.Response(dict(status = 200))
        content = b'{"id":"987","name":"test"}'

    return (res, content)


@contextmanager
def no_env(key):
    value = os.environ.get(key, '')
    os.environ[key] = ''
    yield
    os.environ[key] = value


class TestHTTPGetProvider(unittest.TestCase):
    def test_invalid_provider(self):
        with self.assertRaises(socialauth.InvalidUsage):
            socialauth.http_get_provider('foobar!', '', {}, '')


class TestTwitterProvider(unittest.TestCase):
    def test_no_env(self):
        with no_env('TWITTER_CONSUMER_KEY'):
            with self.assertRaisesRegex(socialauth.Error,
                                        'No TWITTER_CONSUMER_KEY'):
                socialauth.providers.Twitter('', {}, '', '')

        with no_env('TWITTER_CONSUMER_SECRET'):
            with self.assertRaisesRegex(socialauth.Error,
                                        'No TWITTER_CONSUMER_SECRET'):
                socialauth.providers.Twitter('', {}, '', '')

    def test_no_login_start(self):
        with self.assertRaises(socialauth.InvalidUsage):
            socialauth.providers.Twitter('', {}, '', '')

    def get_provider(self, args, token = None):
        return socialauth.http_get_provider(
            'twitter',
            twitter_base_url,
            args,
            'sekret',
            token
        )

    @patch('httplib2.Http.request', mock_valid_requests)
    def test_no_oauth_token_or_verifier(self):
        args = { 'oauth_token': 'Z6eEdO8MOmk394WozF5oKyuAv855l4Mlqo7hhlSLik' }
        with self.assertRaises(socialauth.InvalidUsage):
            self.get_provider(args, 'foobar')

    @patch('httplib2.Http.request', mock_valid_requests)
    def test_no_token_cookie_given(self):
        args = { 'oauth_token': 'foo', 'oauth_verifier': 'foo' }
        with self.assertRaisesRegex(socialauth.Error, 'No token cookie'):
            self.get_provider(args, '')  # Deliberately blank token cookie

    @patch('httplib2.Http.request', mock_valid_requests)
    def test_no_secret_in_web_token(self):
        args  = { 'oauth_token': 'foo', 'oauth_verifier': 'foo' }
        token = jwt.encode({ 'data': { 'type': 'foobar' } }, 'sekret').decode('utf-8')
        with self.assertRaisesRegex(socialauth.Error, 'does not have an oauth_token_secret'):
            self.get_provider(args, token)

    @patch('httplib2.Http.request', mock_valid_requests)
    def test_invalid_web_token(self):
        args  = { 'oauth_token': 'foo', 'oauth_verifier': 'foo' }
        token = jwt.encode({ 'data': { 'type': 'foobar' } }, 'invalid secret').decode('utf-8')
        with self.assertRaisesRegex(socialauth.Error, 'Failed to retrieve'):
            self.get_provider(args, token)

    @patch('httplib2.Http.request', mock_request_with('access_token', 400))
    def test_no_access_token_from_twitter(self):
        args  = { 'oauth_token': 'foo', 'oauth_verifier': 'foo' }
        token = jwt.encode({ 'data': { 'type': 'oauth_token_secret', 'id': 'foo' } }, 'sekret')
        with self.assertRaisesRegex(socialauth.Error, '400 from Twitter'):
            self.get_provider(args, token)

    @patch(
        'httplib2.Http.request',
        mock_request_with(
            'access_token',
            200,
            b'oauth_token=foo&oauth_token_secret=foo&screen_name=test'
        )
    )
    def test_no_user_id_from_twitter(self):
        args  = { 'oauth_token': 'foo', 'oauth_verifier': 'foo' }
        token = jwt.encode({ 'data': { 'type': 'oauth_token_secret', 'id': 'foo' } }, 'sekret')
        with self.assertRaisesRegex(socialauth.Error, 'No user_id'):
            self.get_provider(args, token)

    @patch('httplib2.Http.request', mock_request_with('request_token', 400, ''))
    def test_request_token_failure(self):
        with self.assertRaisesRegex(socialauth.Error, '400 from Twitter'):
            self.get_provider({ 'login': 'start' })

    @patch('httplib2.Http.request', mock_request_with('request_token', 200, b''))
    def test_blank_request_token(self):
        with self.assertRaisesRegex(socialauth.Error, 'No oauth_token'):
            self.get_provider({ 'login': 'start' })

    @patch('httplib2.Http.request', mock_valid_requests)
    def test_flow(self):
        # First leg
        res = self.get_provider({ 'login': 'start' })
        self.assertEqual(res['status'], 302)
        self.assertIn('/oauth/authenticate?oauth_token', res.get('redirect'))

        payload = jwt.decode(res['set_token_cookie'], 'sekret', algorithm = 'HS256')
        self.assertEqual(payload['data']['type'], 'oauth_token_secret')

        # Second leg
        args = {
            'oauth_token': 'Z6eEdO8MOmk394WozF5oKyuAv855l4Mlqo7hhlSLik',
            'oauth_verifier': 'zvq3SztKJphiXzEbUrzt3E7n8WmhZVsx'
        }

        res = self.get_provider(args, res['set_token_cookie'])

        self.assertEqual(res['status'], 200)
        self.assertEqual(res['provider_user_id'], '987')
        self.assertEqual(res['provider_user_name'], 'test')


class TestFacebookProvider(unittest.TestCase):
    def test_no_env(self):
        with no_env('FACEBOOK_APP_ID'):
            with self.assertRaisesRegex(socialauth.Error,
                                        'No FACEBOOK_APP_ID'):
                socialauth.providers.Facebook('', {}, '', '')

        with no_env('FACEBOOK_APP_SECRET'):
            with self.assertRaisesRegex(socialauth.Error,
                                        'No FACEBOOK_APP_SECRET'):
                socialauth.providers.Facebook('', {}, '', '')

    def test_no_login_start(self):
        with self.assertRaises(socialauth.InvalidUsage):
            socialauth.providers.Facebook('', {}, '', '')

    def get_provider(self, args, token = None):
        return socialauth.http_get_provider(
            'facebook',
            facebook_base_url,
            args,
            'sekret',
            None
        )

    @patch('httplib2.Http.request', mock_request_with('access_token', 400))
    def test_failures(self):
        with self.assertRaisesRegex(socialauth.Error, '400 from Facebook'):
            self.get_provider({ 'code': 'foobar' })

    @patch('httplib2.Http.request', mock_request_with('/me', 400))
    def test_graph_me_failure(self):
        with self.assertRaisesRegex(socialauth.Error, '400 from Facebook'):
            self.get_provider({ 'code': 'foobar' })

    @patch(
        'httplib2.Http.request',
        mock_request_with(
            'access_token',
            200,
            b'{"token_type":"bearer","expires_in":5117097}'
        )
    )
    def test_no_access_token_give(self):
        with self.assertRaisesRegex(socialauth.Error, 'No access token'):
            self.get_provider({ 'code': 'foobar' })

    @patch('httplib2.Http.request', mock_request_with('/me', 200, b'{"name":"test"}'))
    def test_graph_me_no_user_id(self):
        with self.assertRaisesRegex(socialauth.Error, 'No user ID'):
            self.get_provider({ 'code': 'foobar' })

    @patch('httplib2.Http.request', mock_valid_requests)
    def test_flow(self):
        res = self.get_provider({ 'login': 'start' })
        self.assertEqual(res['status'], 302)

        # https://developers.facebook.com/docs/facebook-login/manually-build-a-login-flow#login
        args = parse_qs(urlparse(res['redirect'])[4])
        self.assertIn('client_id', args)
        self.assertIn('redirect_uri', args)

        res = self.get_provider({ 'code': 'foo' })
        self.assertEqual(res['status'], 200)
        self.assertEqual(res['provider_user_id'], '987')
        self.assertEqual(res['provider_user_name'], 'test')


if __name__ == '__main__':
    os.environ['TWITTER_CONSUMER_KEY']    = 'foobar'
    os.environ['TWITTER_CONSUMER_SECRET'] = 'foobar'
    os.environ['FACEBOOK_APP_ID']         = '1234'
    os.environ['FACEBOOK_APP_SECRET']     = 'foobar'
    unittest.main()


app.py
# This is a usage example with Flask.
#
# For details and documentation, check out:
# https://socialauth.readthedocs.org/

import os

import dotenv
import jwt
from flask import (
    Flask, request, current_app, redirect,
    jsonify, make_response, abort
)

import socialauth


app = Flask(__name__)


@app.route('/whoami')
def whoami():
    # User IDs are likely protected values, so exposing this value wouldn't
    # be typically recommended in production.
    res = jwt.decode(request.cookies.get('jwt'), current_app.secret_key)
    return res.get('user_id')


# GET /auth/facebook?login=start
# GET /auth/twitter?login=start
@app.route('/auth/<provider>')
def authenticate(provider):
    res = socialauth.http_get_provider(
        provider,
        request.base_url,           # Callback URL
        request.args,               # GET parameters/query string
        current_app.secret_key,
        request.cookies.get('jwt')  # Currently stored token
    )

    if res.get('status') == 302:
        resp = make_response(redirect(res.get('redirect')))
        if res.get('set_token_cookie') is not None:
            resp.set_cookie('jwt', res.get('set_token_cookie'), httponly = True)

        return resp

    if res.get('status') == 200:
        resp = make_response(jsonify({ 'status': 'success' }))

        token = jwt.encode(
            { 'user_id': res.get('provider_user_id') },
            current_app.secret_key,
            algorithm = 'HS256'
        )
        resp.set_cookie('jwt', token, httponly = True)

        return resp

    # Something has gone very wrong. This should not happen.
    abort(400)


if __name__ == '__main__':
    dotenv.load_dotenv('.env')
    app.secret_key = os.urandom(24)
    app.debug = True
    app.run(host = '0.0.0.0')


socialauth/authentication.py
import jwt

import socialauth.providers
from . import InvalidUsage


def validate_provider(provider):
    return provider in (
        'twitter',
        'facebook',
    )


def http_get_provider(provider,
                      request_url, params, token_secret, token_cookie = None):
    '''Handle HTTP GET requests on an authentication endpoint.

    Authentication flow begins when ``params`` has a ``login`` key with a value
    of ``start``. For instance, ``/auth/twitter?login=start``.

    :param str provider: An provider to obtain a user ID from.
    :param str request_url: The authentication endpoint/callback.
    :param dict params: GET parameters from the query string.
    :param str token_secret: An app secret to encode/decode JSON web tokens.
    :param str token_cookie: The current JSON web token, if available.
    :return: A dict containing any of the following possible keys:

        ``status``: an HTTP status code the server should sent

        ``redirect``: where the client should be directed to continue the flow

        ``set_token_cookie``: contains a JSON web token and should be stored by
        the client and passed in the next call.

        ``provider_user_id``: the user ID from the login provider

        ``provider_user_name``: the user name from the login provider
    '''

    if not validate_provider(provider):
        raise InvalidUsage('Provider not supported')

    klass    = getattr(socialauth.providers, provider.capitalize())
    provider = klass(request_url, params, token_secret, token_cookie)
    if provider.status == 302:
        ret = dict(status = 302, redirect = provider.redirect)
        tc  = getattr(provider, 'set_token_cookie', None)
        if tc is not None:
            ret['set_token_cookie'] = tc

        return ret

    if provider.status == 200 and provider.user_id is not None:
        ret = dict(status = 200, provider_user_id = provider.user_id)
        if provider.user_name is not None:
            ret['provider_user_name'] = provider.user_name

        return ret

    raise InvalidUsage('Invalid request')  # pragma: no cover


socialauth/providers/facebook.py
import os
import json
from urllib.parse import quote

import httplib2
import jwt
from socialauth import Error
from socialauth import InvalidUsage


class Facebook:
    def __init__(self, request_url, params, token_secret, token_cookie):
        self.request_url  = request_url
        self.params       = params
        self.token_cookie = token_cookie
        self.token_secret = token_secret

        self.status       = False
        self.access_token = None
        self.user_id      = None
        self.user_name    = None

        self.client_id = os.environ.get('FACEBOOK_APP_ID', None)
        if not self.client_id:
            raise Error('No FACEBOOK_APP_ID environment value')

        self.client_secret = os.environ.get('FACEBOOK_APP_SECRET', None)
        if not self.client_secret:
            raise Error('No FACEBOOK_APP_SECRET environment value')

        self.api_version = os.environ.get('FACEBOOK_GRAPH_API_VERSION', 'v2.5')

        if params.get('login') == 'start':
            self.start()
        elif 'code' in params:
            self.finish()
        else:
            raise InvalidUsage('Invalid request')

    def start(self):
        url = 'https://www.facebook.com/dialog/oauth?client_id={}&redirect_uri={}'
        url = url.format(self.client_id, quote(self.request_url))
        self.status = 302
        self.redirect = url

    def finish(self):
        self.get_access_token()
        self.get_user_information()

        if self.user_id is not None:
            self.status = 200

    def get_access_token(self):
        qs = 'client_id={}&redirect_uri={}&client_secret={}&code={}'
        qs = qs.format(
            self.client_id,
            quote(self.request_url),
            self.client_secret,
            self.params.get('code'))

        url = 'https://graph.facebook.com/{}/oauth/access_token?{}'
        url = url.format(self.api_version, qs)

        resp, content = httplib2.Http().request(url, 'GET')
        if resp.status != 200:
            raise Error('{} from Facebook'.format(resp.status))

        res = json.loads(content.decode('utf-8'))
        access_token = res.get('access_token', None)
        if access_token is None:
            raise Error('No access token from Facebook')

        self.access_token = access_token
        return access_token

    def get_user_information(self):
        url = 'https://graph.facebook.com/me?fields=id,name&access_token={}'
        url = url.format(self.access_token, self.access_token)

        resp, content = httplib2.Http().request(url, 'GET')
        if resp.status != 200:
            raise Error('{} from Facebook'.format(resp.status))

        res = json.loads(content.decode('utf-8'))
        self.user_id = res.get('id', None)
        if not self.user_id:
            raise Error('No user ID from Facebook')

        self.user_name = res.get('name', None)
        return (self.user_id, self.user_name)


socialauth/providers/twitter.py
import os
from urllib.parse import parse_qsl

import oauth2
import jwt

from socialauth import Error
from socialauth import InvalidUsage


class Twitter:
    def __init__(self, request_url, params, token_secret, token_cookie):
        self.request_url  = request_url
        self.params       = params
        self.token_secret = token_secret
        self.token_cookie = token_cookie

        self.status             = False
        self.user_id            = None
        self.user_name          = None

        self.set_token_cookie   = None
        self.oauth_verifier     = params.get('oauth_verifier', None)
        self.oauth_token        = params.get('oauth_token', None)
        self.oauth_token_secret = None

        self.consumer_key = os.environ.get('TWITTER_CONSUMER_KEY', None)
        if not self.consumer_key:
            raise Error('No TWITTER_CONSUMER_KEY environment value')

        self.consumer_secret = os.environ.get('TWITTER_CONSUMER_SECRET', None)
        if not self.consumer_secret:
            raise Error('No TWITTER_CONSUMER_SECRET environment value')

        self.consumer = oauth2.Consumer(
            key       = self.consumer_key,
            secret    = self.consumer_secret)

        if params.get('login') == 'start':
            self.start()
        elif self.oauth_verifier and self.oauth_token:
            self.finish()
        else:
            raise InvalidUsage('Invalid request')

    def start(self):
        self.get_initial_oauth_tokens()
        self.get_redirect_with_token()

    def finish(self):
        self.decode_oauth_token_secret()
        self.get_user_information()

    def decode_oauth_token_secret(self):
        if not self.token_cookie:
            raise Error('No token cookie given')

        try:
            payload = jwt.decode(self.token_cookie,
                                 self.token_secret,
                                 algorithm = 'HS256')
        except:
            raise Error('Failed to retrieve oauth_token_secret from token')

        self.oauth_token_secret = payload.get('data', {}).get('id', None)
        if not self.oauth_token_secret:
            raise Error('Token does not have an oauth_token_secret')

        return self.oauth_token_secret

    def get_user_information(self):
        token = oauth2.Token(self.oauth_token, self.oauth_token_secret)
        token.set_verifier(self.oauth_verifier)
        client = oauth2.Client(self.consumer, token)

        url = 'https://api.twitter.com/oauth/access_token'
        resp, content = client.request(url, 'POST')
        if resp.status != 200:
            raise Error('{} from Twitter'.format(resp.status))

        provider_user = dict(parse_qsl(content.decode('utf-8')))
        if 'user_id' not in provider_user:
            raise Error('No user_id from Twitter')

        self.status    = 200
        self.user_id   = provider_user.get('user_id')
        self.user_name = provider_user.get('screen_name', None)
        return (self.user_id, self.user_name,)

    def get_initial_oauth_tokens(self):
        url = 'https://api.twitter.com/oauth/request_token'
        client = oauth2.Client(self.consumer)
        resp, content = client.request(url, 'GET')
        if resp.status != 200:
            raise Error('{} from Twitter'.format(resp.status))

        oauth_values = dict(parse_qsl(content.decode('utf-8')))
        self.oauth_token        = oauth_values.get('oauth_token')
        self.oauth_token_secret = oauth_values.get('oauth_token_secret')
        if not self.oauth_token or not self.oauth_token_secret:
            raise Error('No oauth_token or oauth_token_secret from Twitter')

        return (self.oauth_token, self.oauth_token_secret,)

    def get_redirect_with_token(self):
        url = 'https://api.twitter.com/oauth/authenticate?oauth_token={}'
        url = url.format(self.oauth_token)

        # Formatted by JSON API spec (jsonapi.org)
        payload = { 'data': {
            'type': 'oauth_token_secret',
            'id': self.oauth_token_secret
        } }

        self.set_token_cookie = jwt.encode(payload,
                                           self.token_secret,
                                           algorithm = 'HS256')
        self.status   = 302
        self.redirect = url
        return (self.redirect, self.set_token_cookie,)
